Multi-Client Implementation

ranj063 opened the pull request:
Multi-Client implementation with 2 clients(for now), one for audio and another dummy client that is used for IPC flood tests.

The work to separate the HDMI pipelines into a separate client is still WIP and will likely be added as a separate PR.

lgirdwood commented on sound/soc/sof/sof-client.c:
```diff
@@ -0,0 +1,242 @@
+
+	return sof_ops(sdev)->drv;
+}
+EXPORT_SYMBOL(sof_client_get_dai_drv);
```
Some of these smaller ones could be static inline.

ranj063 replied:
sure, will change them

ranj063 replied:
@lgirdwood I think I may have spoken too soon here. I actually did consider making them static inline but I ended up not doing it because I wanted to restrict the use of the core ops/definitions ie I want to restrict including the core headers, namely, ops.h and sof-priv.h only to sof-client.c and not the client code. Is there a more elegant way to do this?

lgirdwood replied:
if you want them private thye can still be static inline in ops-priv.h or a similar name.

lgirdwood commented on sound/soc/sof/core.c:
```diff
@@ -50,6 +49,37 @@ static const struct sof_panic_msg panic_msg[] = {
+	if (IS_ERR(client->pdev))
+		goto err;
+
+	list_add(&client->list, &sdev->client_list);
```
Does this need locking ? i.e. are we using this in IRQ handler/thread or in other context sensitive places ?

ranj063 replied:
hmm at the moment, the only place where this is called is by the SOF core at the end of its probe callback.

lgirdwood replied:
What about the IRQ thread ? i.e when we are propagating through the clients when host receives an IPC message (when determining which client gets the IPC).

ranj063 replied:
@lgirdwood this brings the other question in my mind, are clients allowed to register themselves with the SOF core? Or are the clients supposed to be registered by the SOF core itself as has been done for the 2 clients here?

lgirdwood replied:
SOF core exports the registration APIs, its up to clients to then register when they probe().  The mediator will also be a client, clients can also be bespoke.

ranj063 replied:
@lgirdwood makes sense but if I expose an API to register, how would I get a handle to the SOF core device in this API to be able to add the client to the client list?

lgirdwood replied:
We need something like device registration on Linux e.g.
```
/* description of client to help with registration and routing of IPCs */
struct sof_driver {
    const char *name;
    enum sof_drv_class class; // audio, sensing,  etc
    int id;
    // any ops ?
};

struct sof_client * snd_soc_client_register(struct sof_drive *drv);
```

ranj063 replied:
@lgirdwood Sorry maybe I am missing something basic here. In the current implementation, the assumption is that all client devices are children of the SOF core device and they are created and registered by the core itself. But your ask goes a bit further is requiring any device to be able to register with the SOF core. So my question is how does this client device get a pointer to the SOF core device in order to be able to register itself?

lgirdwood replied:
Clients dont initially need a pointer, see above. Clients pass in a "driver" structure and are returned a pointer. This way any client can register without have to modify core.

paulstelian97 replied:
This does mean that there has to be a static that holds the pointer to the SOF core device that the register function has access to, right?

lgirdwood replied:
Yes, it's the same as generic device/driver registration.

ranj063 replied:
@lgirdwood I remember Pierre mentioned (a couple of years ago) that static variables are generally not recommended in kernel drivers. But this is inevitable if we want clients to register themselves with the SOF core. @plbossart do you have an opinion here?

plbossart replied:
why would this be a static? You can devm-allocate a context during the probe of the client module, and keep the sof_client handle there.

ranj063 replied:
@plbossart we're talking about the pointer to SOF core device, sdev, in the above line. When a client register with the SOF core by invoking the client register API, we need add it the list of the client in sdev. But if sdev is not static, how else can we access it?

plbossart replied:
As I mentioned it, you guys are overloading the notion of 'register'.

In the code above, what you are doing is create a child device. Of course the client cannot create such a device on its own, it can only provide a driver for such a child without controlling when the child device is actually created.

If what you want is have the client request a communication channel with the core, then it's a different registration mechanism different from the usual device/driver handling.

plbossart replied:
I might add that I am confused by @lgirdwood statements. In the case where the 'client' will be handling an audio card, so essentially be running a machine driver, then how exactly would this client probe on its own? We have zero mechanisms in ACPI for this at the moment.

ranj063 replied:
> As I mentioned it, you guys are overloading the notion of 'register'.
> 
> In the code above, what you are doing is create a child device. Of course the client cannot create such a device on its own, it can only provide a driver for such a child without controlling when the child device is actually created.
> 
> If what you want is have the client request a communication channel with the core, then it's a different registration mechanism different from the usual device/driver handling.

@plbossart this is the part I'm struggling with in Liam's request. What would that alternate registration mechanism be?

lgirdwood replied:
@plbossart @ranj063 I'm asking you guys to copy the same flows as the Linux driver model. There will probably need to be a ```static LIST_HEAD(client_drivers)``` somewhere in the SOF core (please look at device core, last time I check many years ago it did the same).

1) We register client drivers NOT client devices, the client devices are then ```probed()``` as part of the enumeration. ASoC does this too with components, but there would be no binding or similar enumeration dependencies needed for SOF clients (except that base FW has booted). 
2) We must always pass ```struct sof_client_dev``` in all our APIs. We must not use ```struct device``` as the handle and abuse the private_data between client and core. We can also back cast ```sof_client_dev``` too for any sub APIs that are invented later.

Should I create some example code ?

plbossart replied:
Liam, can you please elaborate on "We register client drivers NOT client devices, the client devices are then probed() as part of the enumeration."

In the case of machine drivers, how exactly would this work? The machine driver is handled as a platform driver, so someone needs to create the device for the machine driver probe to be executed.

Also who's 'we". the SOF core?

ranj063 replied:
@lgirdwood thanks for the clarification. I think I get the idea now except for the part about who registers the client drivers. In the case of ASoC Components, the platform driver registers the component drv. But who would do this in the case of SOF clients?

lgirdwood replied:
I don't think we should have any dependency now on ASoC machine driver probing in order to probe and boot the DSP (this may not be the case today), this way non audio functions do not depend on audio. Can someone confirm if this is the case (sorry, I last deep dived the code about a year ago) ? 

I also think we have to consider two registration uses cases (especially if we move to IPC2.0 with discoverability).

1) Base FW contains five features, all which need a feature driver. SOF core then creates the platform devices for these features on FW boot enumeration data. i.e. FW boot data contains support for audio class, hence SOF infrastructure creates audio platform device (audio driver registration is done by audio class driver).

1.1) SOF core infrastructure probes and boots FW based on PCI/ACPI/DT ID.
1.2) SOF base FW boot meta data contains all supported features and is parsed by SOF core
1.3) SOF core creates platform devices (using existing machine driver names and lookup tables based on machine) for each function.

2) Base FW contains limited functions. Users need to load functions on ad-hoc basis, the sof client driver registers (as part of module insmod flow) and FW module (pointed to by driver structure) is loaded at this point too. This then causes base FW to load the module and emit module boot data (which contains class info) causing the SOF infrastructure to create platform device.

3) rmmod case for 2) above. base FW unloads module, SOF infrastructure destroys platform device and SOD client driver is unregistered.

Does this make sense ? My main thinking here is that we can no longer call the current code the SOF driver but rather the SOF infrastructure.

This would need some minor changes to current machine drivers (platform_driver -> sof_driver),  struct sof_driver could be a container of ```struct platform_driver``` and you would need to create a new module init macro e.g.
```c
/* module_platform_driver() - Helper macro for drivers that don't do
 * anything special in module init/exit.  This eliminates a lot of
 * boilerplate.  Each module may only use this macro once, and
 * calling it replaces module_init() and module_exit()
 */
#define module_platform_driver(__platform_driver) \
	module_driver(__platform_driver, platform_driver_register, \
			platform_driver_unregister)
```
This would need to be ```module_sof_driver```, for example please just grep for ```module_driver``` in include/linux (lots of subsystems are already doing this).

plbossart replied:
@lgirdwood I was ok until the last paragraph. If we create a new type of 'sof_driver', then it means we need to have an sof_device, and given the recent complexities of doing such things with SoundWire we may want to stick to platform devices for now?

lgirdwood replied:
@plbossart it's just a container for platform_driver, you can then add our own ops and stuff in there too, likewise sof_client can be a container of platform_device (again we add our own stuff around it - rather than breaking the platform device platform data for clients).

plbossart replied:
that could work technically indeed if we extend platform drivers, but it's still a pain since we'd modify legacy machine drivers. I am not sure I want to touch all this.

well that said if we take out HDMI from those machine drivers we have to take the scissors anyways. Gah, more work...

ranj063 replied:
> I don't think we should have any dependency now on ASoC machine driver probing in order to probe and boot the DSP (this may not be the case today), this way non audio functions do not depend on audio. Can someone confirm if this is the case (sorry, I last deep dived the code about a year ago) ?

@lgirdwood There's no dependency on machine drv probing today in order to boot the DSP but there is a dependency on finding an ACPI match. But that has all been removed in this PR

lgirdwood commented on sound/soc/sof/sof-audio.c:
```diff
@@ -495,23 +485,179 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+				swidget->widget->id, swidget->comp_id);
+
+			return ret;
+		}
```
We probably want to keep going here and free as many as we can. Lets just log any failures.

ranj063 replied:
@lgirdwood that would probably work if we were to power down the DSP irrespective of the errors during free. But the DSP could be active because of another client and if we do not report the error here during suspend, a resume for this audio client will fail as well.

lgirdwood replied:
I'm thinking here about dependencies or races. i.e. we are draining or we are a component that has dependencies (and cant be freed until all dependent components are freed). 

I the above 2 cases I would expect our components to return -EAGAIN (waiting for draining completion or other event completion) or -EBUSY (for dependencies). So far we have zero components that return these codes, but we should implement and document this in our APIs so that there is no future API change.

ranj063 replied:
@lgirdwood we free the pipeline components only in the case of runtime suspend. This should rule out both the above cases because there should be nothing active at all before runtime suspend kicks in. Does that make sense?

lgirdwood replied:
We have the case when we may be forced to tear down a pipelines (say when a mediator guest crashes).
Some components also may need time to clear any HW state (think DMA buffer draining so may not be immediately ready to free). There is no harm in adding this exception now so that it's there when needed.

lyakh commented on sound/soc/sof/core.c:
```diff
@@ -218,20 +240,13 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+	 * Register audio client.
+	 * This can fail but errors cannot be propagated.
+	 */
+	sof_client_dev_register(sdev, "sof-audio", SOF_CLIENT_AUDIO);
```
@ranj063 I'm not very familiar with the design behind this work, so maybe this has its reasons: do we want the core register various clients or do we want clients register with the core?

ranj063 replied:
@lyakh this is still an open for me as well. In the current design, only the core can register the clients. But this probably is a question that still needs to be addressed.

plbossart replied:
we should be careful about the notion of 'registration', it can mean everything and its logical opposite.

I think we should follow the usual model of having the SOF core create devices based on detected hardware capabilities (HDMI, HDaudio/I2S/SDW codecs) or software options (probes, IPC test, etc).
The client module _init register a driver for a given device and if/when the device is created the driver probes.

We however need to be handle the case where the kernel does not have support for a specific device. In that case the device creation results in nothing, but that's a valid scenario.

ranj063 replied:
@plbossart what about the virtio case? I havent gotten around to reading @lyakh's document yet but how would the guest register with the SOF core?

paulstelian97 replied:
@ranj063 Probably something exposed to qemu or the virtualization software in itself, because the guest cannot really choose anything by itself. Command line parameters of qemu and it would ask for everything needed when initializing the emulated device.

plbossart replied:
right, it's probably more the host vhost parts that would be the SOF client. @lyakh may want to chime in here.

lyakh replied:
My understanding of multi-client support isn't perfect, but aren't clients roughly divided along pipeline lines? I.e. as long as HDMI and SSP pipelines in the topology aren't connected, they can (and will) be separate clients, right? Whereas with virtualisation guests attach to respective host pipelines. So I don't think they can be separate clients? The only possible change might be the linking method because now I might also need a client ID, or maybe it won't be needed either, if component IDs remain unique across all clients.

plbossart replied:
Yes, the intent was to have an SSP+DMIC card, and an HDMI card. both modeled as clients.

But indeed the virtualization is somewhat orthogonal to the notion of client, since the connections happen at the pipeline level. So you would probably have a single client for all of virtualization uses.

But that brings one open: do we expose HDMI pipelines to the guest? If yes, we'd need means for those pipelines to be supported with jack detection, and we'd need mixers on the HDMI path, which we currently don't.

one of those things where by asking a question you get more work in return :-)

ranj063 replied:
> My understanding of multi-client support isn't perfect, but aren't clients roughly divided along pipeline lines? I.e. as long as HDMI and SSP pipelines in the topology aren't connected, they can (and will) be separate clients, right? Whereas with virtualisation guests attach to respective host pipelines. So I don't think they can be separate clients? The only possible change might be the linking method because now I might also need a client ID, or maybe it won't be needed either, if component IDs remain unique across all clients.

@lyakh I have kept the component IDs unique across all client consciously. I dont currently have the notion of a client ID and besides, the FW doesnt quite partition components by clients. So creating components with the same ID across clients just caused errors in my case.

paulstelian97 replied:
My two cents, is it not possible/reasonable to keep a mapping on the kernel side between client component IDs (which should also include client ID etc) and firmware component IDs? So that unique IDs can be allocated for the firmware even if they will be identical across clients.

ranj063 replied:
@paulstelian97 it is entirely possible to maintain this look-up table between the kernel and firmware. But do you think its essential? The SOF core maintains a linearly incrementing component ID that all the clients can simply query for and use it. What would be the added benefit of having the extra logic to look up the FW component IDs from the client component IDs?

paulstelian97 replied:
Besides conflicts when two clients request the same ID nothing really. Clients have statically selected IDs; the firmware may not need to know the specific ID the client sees but the unique, dynamically allocated ID that is selected via such a mechanism. All IPC would have this ID translated on need so the firmware isn't even made aware of multiple clients (so this would thus be a kernel-only change).

I don't consider this essential though, just one way to go with it. You can feel free to ignore if another alternative is better.

ranj063 replied:
@paulstelian97 this is a good point about conflicting comp ID requests from clients simultaneously. Adding a lock/mutex when might solve the problem

paulstelian97 replied:
@ranj063 So you say that the ID namespace itself should still be shared? My idea was that each client had his own namespace of IDs for components and pipelines so they don't touch the components of other clients. Or you consider that sharing might be useful? (the mapping which I mention can allow a component or pipeline to show up in multiple such namespaces, to be shared explicitly)

ranj063 replied:
@paulstelian97 in the current implementation, clients maintain a list of widgets based on the topology that is loaded when the client devices are probed. So, if they want to do perform operations on those widgets, they can look up widgets' comp ID from their own list of widgets. And this list is maintained in the private data for each client. So, it is not possible for a client to tamper with a component that belongs to another client because there would really be no way for it to look up a component from another client's widget list. Does that make sense?

paulstelian97 replied:
@ranj063 I think so, so the clients know the actual component IDs (without namespacing) but can't work with other components that don't belong to them. Then there must be some dynamic allocation of component IDs (topology doesn't hardcode the actual IDs used at runtime)? Otherwise we will see clashes. I haven't fully grasped the topology loading code so if this is there just point it out and I will be okay with that.

ranj063 replied:
@paulstelian97 here's where the comp ID's are assigned while parsing topology. The next_comp_id field of sdev is a linearly increasing value with each widget that is created in the DSP FW. And this is maintained by the core. So lets say you have 2 clients, one for I2S pipelines with N widgets and one for HDMI pipelines with M widgets. The valid comp ID's for the I2S client would run from 0 to N-1 and the valid comp ID's for the HDMI client would run from N -> N + M - 1. And so typically, when a client wants to perform operations on its components, it would look it up from its own widget list and the components in the other clients' widget list would be off limits.
https://github.com/thesofproject/linux/blob/4cef169b7bf4c6033e8e478921349ae23e5ab300/sound/soc/sof/topology.c#L2217

paulstelian97 replied:
Oh great, then my issue is invalid :) Now my only thing is if there are components that may be explicitly shared between clients. Is this not supported or? (again my understanding is pretty limited, just curious). You would bridge the connections between clients using custom port components or there is some sharing?

ranj063 replied:
@paulstelian97 as I understand, if there are shared components between pipelines, it is better to keep them in the same client thereby meaning that they also remain part of the same card. And this is the reason why we are opting to keep the DMIC pipelines clubbed together with the codec pipelines so that when there is a need to add features such as echo cancellation, we are not dealing with multiple cards.

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,10 +153,11 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 /* Send DAI_CONFIG IPC to the DAI that matches the dai_name and direction */
 static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
-			       const char *dai_name, int channel, int dir)
+			       const char *dai_name, int channel, int dir,
```
looks like we don't need to pass `dai->name` additionally then?

lyakh commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -825,15 +825,31 @@ void hda_dsp_d0i3_work(struct work_struct *work)
 	struct sof_dsp_power_state target_state;
+	struct snd_sof_client *client;
+	struct device *dev;
+	bool remain;
```
I think the compiler should warn you about uninitialised `remain` and in this case it would be right

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -20,6 +21,9 @@
 {
+	struct snd_sof_client *client;
+	struct device *dev;
+
```
superfluous empty line

lyakh commented on sound/soc/sof/pm.c:
```diff
@@ -31,13 +35,19 @@ static u32 snd_sof_dsp_power_target(struct snd_sof_dev *sdev)
+			dev = &client->pdev->dev;
+			if (snd_sof_stream_suspend_ignored(dev)) {
+				target_dsp_state = SOF_DSP_PM_D0;
+				return target_dsp_state;
```
```
if (snd_sof_stream_suspend_ignored(dev))
	return SOF_DSP_PM_D0;
```
and you can actually remove the `target_dsp_state` variable here and `return` immediately wherever it's assigned. I think that would make a clean function, whose sole purpose is to return a target power state. That would be my preferred solution. Or if you want to keep a single return point, which is what this function seems to prefer by sacrificing some performance
```
target_dsp_state = SOF_DSP_PM_D3;
list_for_each_entry(client, &sdev->client_list, list) {
	dev = &client->pdev->dev;
	if (snd_sof_stream_suspend_ignored(dev)) {
		target_dsp_state = SOF_DSP_PM_D0;
		break;
	}
}
break;
```

lyakh commented on sound/soc/sof/sof-audio.c:
```diff
@@ -8,16 +8,20 @@
 {
-	struct snd_sof_audio_data *audio_data = sdev->sof_audio_data;
+	struct snd_sof_audio_data *audio_data = sof_get_client_data(dev);
+
```
superfluous empty line

lyakh commented on sound/soc/sof/sof-audio.c:
```diff
@@ -495,23 +485,179 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+	int ret = 0;
+
+	list_for_each_entry_reverse(swidget, &audio_data->widget_list, list) {
+		memset(&ipc_free, 0, sizeof(ipc_free));
```
put this 3 lines down - after skipping, but in fact I'd just remove it - you're assigning all the `ipc_free` fields below.

lyakh commented on sound/soc/sof/sof-audio.c:
```diff
@@ -495,23 +485,179 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
+{
+	struct snd_sof_audio_data *audio_data =	sof_get_client_data(dev);
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_free ipc_free;
```
I would put this definition inside the loop to make its scope explicit.

lyakh commented on sound/soc/sof/sof-ipc-test.c:
```diff
@@ -0,0 +1,299 @@
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/pm_runtime.h>
+#include <linux/debugfs.h>
```
it's almost alphabetically sorted, which is nice, just move debugfs.h to its place too

fredoh9 commented:
SOFCI TEST

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1089,6 +1090,7 @@ static bool link_slaves_found(struct snd_sof_dev *sdev,
 
 static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 {
+	struct snd_sof_audio_data *audio_data = sdev->sof_audio_data;
```
can we try to deal with SoundWire with dedicated patches, so that submitting upstream is less painful, depending on which of multi-client or SoundWire is merged first?

plbossart commented on sound/soc/sof/Makefile:
```diff
@@ -3,7 +3,7 @@
 
 snd-sof-objs := core.o ops.o loader.o ipc.o pcm.o pm.o debug.o topology.o\
-		control.o trace.o utils.o sof-audio.o
+		control.o trace.o utils.o sof-audio.o sof-client.o
```
should the client be in a separate module with namespaces to force a clean split?

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -50,6 +49,40 @@ static const struct sof_panic_msg panic_msg[] = {
+	if (IS_ERR(client->pdev))
+		goto err;
+
+	list_add(&client->list, &sdev->client_list);
```
we have the recurring problem with platform_devices that you don't know when the driver probe finishes. It's not a good assumption to think that there is a synchronous behavior, so we'd likely need some sort of probe_complete callback before adding the client to the list and claiming victory.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -246,8 +261,6 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	return 0;
 
-fw_trace_err:
-	snd_sof_free_trace(sdev);
```
I think the trace does not belong in the clients but in the core.
Not sure why it's removed here and not in snd_sof_device_remove() below.

ranj063 replied:
@plbossart its removed here because previously machine register could fail but there's no failure after its moved to the audio client. But it still needs to be freed when the SOF device is removed.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -323,8 +322,8 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
-		ret = hda_link_config_ipc(hda_stream, dai->name,
-					  DMA_CHAN_INVALID, substream->stream);
+		ret = hda_link_config_ipc(hda_stream, DMA_CHAN_INVALID,
+					  substream->stream, dai);
```
probably best to keep the arguments in the same order for reviewers, i.e just a dai->name to dai change

plbossart commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -825,15 +825,31 @@ void hda_dsp_d0i3_work(struct work_struct *work)
+			continue;
+
+		if (!snd_sof_dsp_only_d0i3_compatible_stream_active(dev))
+			remain = true;
```
I am not a big fan of this, this looks super-racy.
We'd probably need some sort of locking, but it'd be even better to have notifications with a mutex that changes the internal state.

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -148,13 +148,21 @@ snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
+		spcm = snd_sof_find_spcm_dai(&client->pdev->dev, rtd);
+		if (spcm)
+			break;
+	}
```
the pointer callback is quite latency sensitive, we should probably find a means to match dai id and spcm on startup and avoid this search every period (when IPC updates are requested)

ranj063 replied:
@plbossart yes I could do that to reduce the recursive search here but we'd still have one loop to find the spcm matching the dai id from the list of open streams in case there are more than one. Is that ok?

plbossart replied:
Sorry I don't get the answer. I suggested this loop be removed and you replied that it can be removed but is still needed. what I am missing?

ranj063 replied:
@plbossart there are 2 loops here. The outer loops for the list of clients and the inner loop in snd_sof_find_spcm_dai(). The inner loop has always been there, its only the outer loop that is an addition in this PR because now we have multiple clients, we have to search them all to find the spcm matcing the dai ID. 

Maybe I didnt understand your suggestion correctly. Are you suggesting to remove just the outer loop to iterate through the client list? or both the loops? In either case, I dont think we can remove the loop entirely because even if saved the spcm at startup, in the case of multiple streams open, we'd need to find one which one to choose here.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -754,14 +754,6 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
-		dev_err(sdev->dev, "error: failed to create DMIC device\n");
-		return PTR_ERR(hdev->dmic_dev);
-	}
-
```
do we want to move the dmic device creation to the client?
This would force the client driver to check what platform it runs on, when this is not needed when you create the dmic device from a platform-specific probe.
If we want a generic client then we should keep this here?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -433,13 +433,22 @@ static void ipc_trace_message(struct snd_sof_dev *sdev, u32 msg_id)
+					      &direction);
+		if (spcm)
+			break;
+	}
```
same comment as above for dais, can we skip the search and somehow keep track of the spcm on startup?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -463,13 +472,22 @@ static void ipc_period_elapsed(struct snd_sof_dev *sdev, u32 msg_id)
+					      &direction);
+		if (spcm)
+			break;
+	}
```
same here.

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -68,19 +68,33 @@ static inline int snd_sof_dsp_reset(struct snd_sof_dev *sdev)
+
+		/* update enabled cores mask if the op is successful */
+		if (ret >= 0)
+			sdev->enabled_cores_mask |= core_mask;
```
is this a bad merge/cherry-pick? What does this have to do with multi-client?

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -390,38 +404,38 @@ snd_sof_pcm_platform_pointer(struct snd_sof_dev *sdev,
 		return sof_ops(sdev)->pcm_pointer(sdev, substream);
 
-	return 0;
+	return -ENOTSUPP;
```
why?

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -138,7 +135,7 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
 	 */
 	if (spcm->prepared[substream->stream]) {
-		ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
+		ret = sof_pcm_dsp_pcm_free(substream, component->dev, spcm);
```
can we move the sdev to component->dev changes in another patch?

plbossart commented on sound/soc/sof/trace.c:
```diff
@@ -11,6 +11,7 @@
 #include <linux/debugfs.h>
 #include <linux/sched/signal.h>
 #include "sof-priv.h"
+#include "sof-audio.h"
```
this is a massive commit that needs to be broken up to help reviews

plbossart commented on sound/soc/sof/sof-ipc-test.c:
```diff
@@ -0,0 +1,299 @@
+
+static ssize_t sof_ipc_dfsentry_write(struct file *file,
+				      const char __user *buffer,
+				      size_t count, loff_t *ppos)
```
careful here, all debugfs stuff needs to be GPLv2.0, not BSD

ranj063 replied:
@plbossart Sorry i didnt understand your point here. The current debug.c file is dual licensed currently and I'm just moving the code from there to a separate client. Are you saying that it was wrong originally?

plbossart replied:
possibly. We do know that all debugfs stuff uses GPLv2 exports, e.g.

debug.c:EXPORT_SYMBOL_GPL(snd_sof_debugfs_io_item);
debug.c:EXPORT_SYMBOL_GPL(snd_sof_debugfs_buf_item);
debug.c:EXPORT_SYMBOL_GPL(snd_sof_dbg_init);
debug.c:EXPORT_SYMBOL_GPL(snd_sof_free_debug);

IANAL, but having such symbols in a dual-licensed file looks odd at best.

That said, I see similar issues in other parts of the kernel so probably best to check.

ranj063 replied:
I see then I suppose both the debug.c and the new client can only be GPLv2.0? Let me check with Greg on that.

plbossart replied:
let's check with Josh first.

ranj063 replied:
oops, I just sent an email to Greg. Hope thats not a problem but I did notice that most files in the kernel that used debugfs are GLP2.0 only

ranj063 commented:
@plbossart @lyakh @lgirdwood I've updated the series now. I've addressed most of the feedback except for the following (probably a few more that I may have missed):
1. There is still one large commit. I am still working on how to break it up further but its not an easy task.
2. Client devices are still children of SOF core devices. Registering devices that are not children probably needs to addressed separately.

ranj063 commented:
SOFCI TEST

lgirdwood commented:
My only concern is that this API should look exactly like other kernel APIs that do the same thing i.e.
```c
/* SOF client driver */
struct sof_client_driver {
    char *name;
    enum sof_client_class class;
    struct ops * ops;
};

/* SOF client device */
struct sof_client_dev {
    struct device *dev;
    struct sof_client_driver *drv;
    int client_id; /* used for IPC routing */ 
    ....
};

/* registration would look like */
struct sof_client_dev *sof_client_register(struct sof_client_driver *drv);
void sof_client_unregister(sof_client_dev *sdev);

/* all API would then all llok like */
int sof_client_do_this(struct sof_client_dev *dev, int this);
```
This way no core changes are required to add a new client. Ideally we want to be able to use a stock kernel and base FW with new clients being added as modules to both driver and FW (without any core changes).

ranj063 commented:
> My only concern is that this API should look exactly like other kernel APIs that do the same thing i.e.
> 
> ```c
> /* SOF client driver */
> struct sof_client_driver {
>     char *name;
>     enum sof_client_class class;
>     struct ops * ops;
> };
> 
> /* SOF client device */
> struct sof_client_dev {
>     struct device *dev;
>     struct sof_client_driver *drv;
>     int client_id; /* used for IPC routing */ 
>     ....
> };
> 
> /* registration would look like */
> struct sof_client_dev *sof_client_register(struct sof_client_driver *drv);
> void sof_client_unregister(sof_client_dev *sdev);
> 
> /* all API would then all llok like */
> int sof_client_do_this(struct sof_client_dev *dev, int this);
> ```
> 
> This way no core changes are required to add a new client. Ideally we want to be able to use a stock kernel and base FW with new clients being added as modules to both driver and FW (without any core changes).

@lgirdwood Thanks. I will fix these. But may I ask what the ops in the sof_client_driver struct is for?

fredoh9 commented:
SOFCI TEST

lgirdwood commented:
@ranj063 ops will be ```probe() remove()``` if you want them (it would mean changing the registration a little) alongside any generic ops/callbacks clients need.

ranj063 commented:
SOFCI TEST

ranj063 commented:
closing this one. There's a bit of a change in the direction for the sof client drv model. will open a new one soon
